File: .temp/flutter_app/lib/data/local/dao/tag_dao.dart
import 'package:drift/drift.dart';
import '../database.dart';

part 'tag_dao.g.dart';

@DriftAccessor(tables: [Tags])
class TagDao extends DatabaseAccessor<AppDatabase> with _$TagDaoMixin {
  TagDao(super.db);

  Stream<List<TagEntity>> getAllTags() {
    return (select(tags)
          ..orderBy([
            (t) => OrderingTerm(expression: t.orderIndex, mode: OrderingMode.asc),
            (t) => OrderingTerm(expression: t.id, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Future<int> insertTag(Insertable<TagEntity> tag) {
    return into(tags).insert(tag, mode: InsertMode.replace);
  }

  Future<TagEntity?> getTagByName(String name) {
    return (select(tags)..where((t) => t.name.equals(name))).getSingleOrNull();
  }

  Future<bool> updateTag(Insertable<TagEntity> tag) {
    return update(tags).replace(tag);
  }

  Future<void> updateTags(List<Insertable<TagEntity>> tagList) async {
    await batch((batch) {
      batch.replaceAll(tags, tagList);
    });
  }

  Future<int> deleteTag(Insertable<TagEntity> tag) {
    return delete(tags).delete(tag);
  }
}

File: shared/src/commonMain/kotlin/com/sans/finance/data/local/dao/TagDao.kt
package com.sans.finance.data.local.dao

import androidx.room3.Dao
import androidx.room3.Delete
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import androidx.room3.Update
import com.sans.finance.data.local.entity.TagEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface TagDao {
    @Query("SELECT * FROM tags ORDER BY orderIndex ASC, id ASC")
    fun getAllTags(): Flow<List<TagEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertTag(tag: TagEntity): Long

    @Query("SELECT * FROM tags WHERE name = :name LIMIT 1")
    suspend fun getTagByName(name: String): TagEntity?

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateTag(tag: TagEntity)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateTags(tags: List<TagEntity>)

    @Delete
    suspend fun deleteTag(tag: TagEntity)

    @Query("DELETE FROM tags WHERE id NOT IN (SELECT DISTINCT tagId FROM expense_tag_ref)")
    suspend fun deleteOrphanedTags()
}
